File: HMW6/homework6/Form1.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Globalization;
using System.Linq;
using System.Windows.Forms;
using Microsoft.VisualBasic.FileIO;

namespace homework6
{
    public partial class Form1 : Form
    {
        Bitmap b;
        Graphics g;
        Pen PenTrajectory = new Pen(Color.DarkGreen, 1);

        Bitmap bIsto;
        Graphics gIsto;
        Boolean boolWeight = true;

        public Form1()
        {
            InitializeComponent();
        }

        private void richTextBox1_TextChanged(object sender, EventArgs e)
        {

        }

        private void button1_Click(object sender, EventArgs e)
        {
            this.b = new Bitmap(this.pictureBox1.Width, this.pictureBox1.Height);
            this.g = Graphics.FromImage(b);
            this.g.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality;
            this.g.Clear(Color.White);
            this.richTextBox1.Clear();

            Rectangle VirtualWindow = new Rectangle(0, 0, this.b.Width - 1, this.b.Height - 1);
            g.DrawRectangle(Pens.DarkSlateGray, VirtualWindow);
            int index = 1;
            Random r = new Random();
            if (boolWeight) index = 1; //1 is weigth 2 is heigth
            else index = 2;

            int trialsnumber = 300, samplesize = 20; 
            var valuestring = new List<double>();

            double minX = 0;
            double maxX = samplesize;
            double maxValue = 0;
            double minValue = 0;
            double minY = minValue;
            double maxY = maxValue;

            using (TextFieldParser parser = new TextFieldParser(@"./../../weight-height.csv"))
            {
                parser.TextFieldType = FieldType.Delimited;
                parser.SetDelimiters(",");
                while (!parser.EndOfData)
                {
                    string[] fields = parser.ReadFields();
                    double key = double.Parse(fields[index], CultureInfo.InvariantCulture);
                    valuestring.Add(key);
                }
            }

            int length = valuestring.Count;
            List<double> avglist = new List<double>();

            for (int i = 0; i < trialsnumber; i++)
            {
                List<Point> Punti = new List<Point>();
                List<int> selectorList = new List<int>();


                double tempmean = 0;
                double mean = 0;

                for (int j = 0; j <= samplesize; j++)
                {
                    int selector = r.Next(length);
                    selectorList.Add(selector);

                    double value = valuestring[selector];
                    
                    mean = (mean + value);
                    tempmean = mean / (j+1);


                    if (minValue == 0 || tempmean < minValue)
                    {
                        minValue = tempmean;
                    }

                    if (minValue == 0 || tempmean > maxValue)
                    {
                        maxValue = tempmean;
                    }

                    minY = minValue;
                    maxY = maxValue;

                }
                avglist.Add(tempmean);

                tempmean = 0;
                mean = 0;

                for (int j = 0; j < selectorList.Count; j++)
                {
                    int random = selectorList[j];
                    double value = valuestring[random];

                    mean = (mean + value);
                    tempmean = mean / (j + 1);

                    int xDevice = (int)(FromXRealToXVirtual(j, minX, maxX, VirtualWindow.Left, VirtualWindow.Width));
                    int YDevice = (int)(FromYRealToYVirtual(tempmean, minY, maxY, VirtualWindow.Top, VirtualWindow.Height));
                    Punti.Add(new Point(xDevice, YDevice));
                }

                

                //g.DrawLines(PenTrajectory, Punti.ToArray());

            }
            int n = 1;
                foreach (double j in avglist)
                {
                    this.richTextBox1.AppendText("average of sample" + n.ToString() + ": " + j.ToString() + "\n");
                    n++;
                }

            this.pictureBox1.Image = b;

            this.bIsto = new Bitmap(this.pictureBox1.Width, this.pictureBox1.Height);
            this.gIsto = Graphics.FromImage(bIsto);

            this.gIsto.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality;
            this.gIsto.Clear(Color.White);

            Rectangle VirtualWindow2 = new Rectangle(0, 0, this.bIsto.Width - 1, this.bIsto.Height - 1);
            gIsto.DrawRectangle(Pens.Black, VirtualWindow2);

            double minAvg = avglist.Min();
            double maxAvg = avglist.Max();
            double delta = maxAvg - minAvg;
            double nintervals = 20;
            double intervalsSize = delta / nintervals;
            Dictionary<double, int> istoDict = new Dictionary<double, int>();

            double tempValue = minAvg;
            for (int i = 0; i < nintervals; i++)
            {
                istoDict[tempValue] = 0;
                tempValue = tempValue + intervalsSize;
            }

            int total = 0;

            foreach (double value in avglist)
            {
                foreach (double key in istoDict.Keys)
                {
                    if (value < key + intervalsSize)
                    {
                        total++;
                        istoDict[key] += 1;
                        break;
                    }
                }
            }

            gIsto.TranslateTransform(0, this.bIsto.Height);
            gIsto.ScaleTransform(1, -1);

            int idIsto = 0;
            int widthIsto = (int)(this.bIsto.Width / nintervals);
            foreach (double key in istoDict.Keys)
            {
                int newHeight = istoDict[key] * this.bIsto.Height / total;
                int newX = (widthIsto * idIsto) + 1;
                Rectangle isto = new Rectangle(newX, 0, widthIsto, newHeight);
                idIsto++;

                int nextWidthIstogram = (int)(widthIsto * idIsto * 1);
                
                SolidBrush the_brush = new SolidBrush(Color.Red);
                gIsto.DrawRectangle(Pens.Black, isto);
                gIsto.FillRectangle(the_brush, isto);
            }
            this.pictureBox1.Image = bIsto;


        }

        private void pictureBox1_Click(object sender, EventArgs e)
        {

        }

            public double FromXRealToXVirtual(double X, double minX, double maxX, int Left, int W)
            {
                if ((maxX - minX) == 0)
                    return 0;

                return Left + W * (X - minX) / (maxX - minX);
            }

            public double FromYRealToYVirtual(double Y, double minY, double maxY, int Top, int H)
            {
                if ((maxY - minY) == 0)
                    return 0;

                return Top + H - H * (Y - minY) / (maxY - minY);
            }

        private void pictureBox2_Click(object sender, EventArgs e)
        {

        }

        private void button2_Click(object sender, EventArgs e)
        {
            this.b = new Bitmap(this.pictureBox2.Width, this.pictureBox2.Height);
            this.g = Graphics.FromImage(b);
            this.g.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality;
            this.g.Clear(Color.White);
            this.richTextBox2.Clear();


            Rectangle VirtualWindow = new Rectangle(0, 0, this.b.Width - 1, this.b.Height - 1);
            g.DrawRectangle(Pens.DarkSlateGray, VirtualWindow);

            Random r = new Random();
            int index;

            if (boolWeight) index = 1; //1 is weigth 2 is heigth
            else index = 2;
            int trialsnumber = 300, samplesize = 20; 
            var valuestring = new List<double>();

            double minX = 0;
            double maxX = samplesize;
            double maxValue = 0;
            double minValue = 0;
            double minY = minValue;
            double maxY = maxValue;

            using (TextFieldParser parser = new TextFieldParser(@"./../../weight-height.csv"))
            {
                parser.TextFieldType = FieldType.Delimited;
                parser.SetDelimiters(",");
                while (!parser.EndOfData)
                {
                    string[] fields = parser.ReadFields();
                    double key = double.Parse(fields[index], CultureInfo.InvariantCulture);
                    valuestring.Add(key);
                }
            }

            int length = valuestring.Count;
            List<double> variancelist = new List<double>();

            for (int i = 0; i < trialsnumber; i++)
            {
                List<Point> Punti = new List<Point>();
                List<int> selectorList = new List<int>();


                double tempmean = 0;
                double mean = 0;
                double variance = 0.0;

                for (int j = 0; j <= samplesize; j++)
                {
                    int selector = r.Next(length);
                    selectorList.Add(selector);

                    double value = valuestring[selector];

                    mean = (mean + value);
                    tempmean = mean / (j + 1);

                    variance += Math.Pow(value - tempmean, 2.0);
                    variance= (variance/(j + 1))*3;

                    if (minValue == 0 || variance < minValue)
                    {
                        minValue = variance;
                    }

                    if (minValue == 0 || variance > maxValue)
                    {
                        maxValue = variance;
                    }

                    minY = minValue;
                    maxY = maxValue;

                }
                variancelist.Add(variance);


                tempmean = 0;
                mean = 0;
                double variance1 = 0.0;

                for (int j = 0; j < selectorList.Count; j++)
                {
                    int itm = selectorList[j];
                    double value = valuestring[itm];

                    mean = (mean + value);
                    tempmean = mean / (j + 1);

                    variance1 += Math.Pow(value - tempmean, 2.0);
                    variance1 = variance1 / (j + 1);

                    int xDevice = (int)(FromXRealToXVirtual(j, minX, maxX, VirtualWindow.Left, VirtualWindow.Width));
                    int YDevice = (int)(FromYRealToYVirtual(variance1, minY, maxY, VirtualWindow.Top, VirtualWindow.Height));
                    Punti.Add(new Point(xDevice, YDevice));
                }
                
                //g.DrawLines(PenTrajectory, Punti.ToArray());



            }
            int n = 1;
                foreach (double j in variancelist)
                {
                    this.richTextBox2.AppendText("variance of sample" + n.ToString() + ": " + j.ToString() + "\n");
                    n++;
                }
            this.pictureBox1.Image = b;

            this.bIsto = new Bitmap(this.pictureBox1.Width, this.pictureBox1.Height);
            this.gIsto = Graphics.FromImage(bIsto);

            this.gIsto.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality;
            this.gIsto.Clear(Color.White);

            Rectangle VirtualWindow2 = new Rectangle(0, 0, this.bIsto.Width - 1, this.bIsto.Height - 1);
            gIsto.DrawRectangle(Pens.Black, VirtualWindow2);

            double minAvg = variancelist.Min();
            double maxAvg = variancelist.Max();
            double delta = maxAvg - minAvg;
            double nintervals = 20;
            double intervalsSize = delta / nintervals;

            Dictionary<double, int> istoDict = new Dictionary<double, int>();

            double tempValue = minAvg;
            for (int i = 0; i < nintervals; i++)
            {
                istoDict[tempValue] = 0;
                tempValue = tempValue + intervalsSize;
            }

            int total = 0;

            foreach (double value in variancelist)
            {
                foreach (double key in istoDict.Keys)
                {
                    if (value < key + intervalsSize)
                    {
                        total++;
                        istoDict[key] += 1;
                        break;
                    }
                }
            }

            gIsto.TranslateTransform(0, this.bIsto.Height);
            gIsto.ScaleTransform(1, -1);

            int idIsto = 0;
            int widthIsto = (int)(this.bIsto.Width / nintervals);
            foreach (double key in istoDict.Keys)
            {
                int newHeight = istoDict[key] * this.bIsto.Height / total;
                int newX = (widthIsto * idIsto) + 1;
                Rectangle isto = new Rectangle(newX, 0, widthIsto, newHeight);
                idIsto++;

                int nextWidthIstogram = (int)(widthIsto * idIsto * 1);

                SolidBrush the_brush = new SolidBrush(Color.Red);
                gIsto.DrawRectangle(Pens.Black, isto);
                gIsto.FillRectangle(the_brush, isto);
            }
            this.pictureBox2.Image = bIsto;
 
        }

        private void richTextBox1_TextChanged_1(object sender, EventArgs e)
        {

        }

        private void richTextBox2_TextChanged(object sender, EventArgs e)
        {

        }

        private void radioButton1_CheckedChanged(object sender, EventArgs e)
        {

        }

        private void radioButton2_CheckedChanged(object sender, EventArgs e)
        {
            if (boolWeight) boolWeight = false;
            else boolWeight = true;
        }
    }
}

File: Applicazione Tesina/RandomWalkGenerator/RandomWalkGenerator/Form1.vb
Public Class Form1

    Public b, b2 As Bitmap
    Public g, g2 As Graphics
    Public r As New Random
    Private Sub PictureBox1_Click(sender As Object, e As EventArgs) Handles PictureBox1.Click

    End Sub

    Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click
        Me.b = New Bitmap(Me.PictureBox1.Width, Me.PictureBox1.Height)
        Me.g = Graphics.FromImage(b)
        Me.g.SmoothingMode = Drawing2D.SmoothingMode.HighQuality

        Me.b2 = New Bitmap(Me.PictureBox1.Width, Me.PictureBox2.Height)
        Me.g2 = Graphics.FromImage(b2)
        Me.g2.SmoothingMode = Drawing2D.SmoothingMode.HighQuality

        Me.g.Clear(Color.White)
        Me.g2.Clear(Color.White)

        Dim TrialsCount As Integer = Me.NumericUpDown1.Value
        Dim NumerOfTrajectories As Integer = Me.NumericUpDown2.Value
        Dim SuccessProbability As Double = 0.5

        Dim minX As Double = 0
        Dim maxX As Double = TrialsCount
        Dim minY As Double = 0
        Dim maxY As Double = TrialsCount * 2
        Dim VirtualWindow As New Rectangle(0, 0, Me.b.Width, Me.b.Height)
        g.DrawRectangle(Pens.DarkSlateGray, VirtualWindow)

        Dim xDevice As Integer
        Dim dictY As New Dictionary(Of Integer, Integer)

        For i As Integer = 0 To NumerOfTrajectories

            Dim Punti As New List(Of Point)
            Dim Punti2 As New List(Of Point)

            Dim walkupdown As Double = 0
            Dim YDevice As Integer


            For X As Integer = 0 To TrialsCount
                Dim Uniform As Double = r.NextDouble()
                If Uniform < SuccessProbability Then
                    walkupdown += 3
                    If walkupdown > maxY Then
                        maxY = walkupdown
                    End If
                Else
                    walkupdown -= 3

                End If
                xDevice = FromXRealToXVirtual(X, minX, maxX, VirtualWindow.Left, VirtualWindow.Width)

                YDevice = FromYRealToYVirtual(walkupdown, minY, maxY, VirtualWindow.Top, VirtualWindow.Height / 2)
                Punti.Add(New Point(xDevice, YDevice))

            Next

            If dictY.ContainsKey(YDevice) Then
                dictY(YDevice) += 1
            Else
                dictY.Add(YDevice, 1)
            End If

            Dim rng As New Random()
            For x = 1 To 150
                Using pen = New Pen(Color.FromArgb(rng.Next(256), rng.Next(256), rng.Next(256)))
                    g.DrawLines(pen, Punti.ToArray)
                End Using
            Next x

        Next
        Dim rectwidthy As Double
        Dim indexy, scalahy As Double
        scalahy = dictY.Values.Max / Me.PictureBox2.Width
        indexy = 0
        'rectwidthy = Me.PictureBox3.Width / dictY.Count

        For Each item In dictY

            Dim rect As New Rectangle(0, item.Key, item.Value / scalahy, 3)
            Dim the_brush As New SolidBrush(Color.Pink)
            Dim the_pen As New Pen(Color.Black, 0)
            g2.FillRectangle(the_brush, rect)
            g2.DrawRectangle(the_pen, rect)
            indexy = indexy + rectwidthy

        Next
        Me.PictureBox1.Image = b
        Me.PictureBox2.Image = b2
    End Sub

    Function FromXRealToXVirtual(X As Double,
                                 minX As Double, maxX As Double,
                                 Left As Integer, W As Integer) As Integer

        If (maxX - minX) = 0 Then
            Return 0
        End If

        Return Left + W * (X - minX) / (maxX - minX)

    End Function

    Function FromYRealToYVirtual(Y As Double,
                                minY As Double, maxY As Double,
                                Top As Integer, H As Integer) As Integer

        If (maxY - minY) = 0 Then
            Return 0
        End If

        Return Top + H - H * (Y - minY) / (maxY - minY)

    End Function

    Private Sub NumericUpDown1_ValueChanged(sender As Object, e As EventArgs) Handles NumericUpDown1.ValueChanged

    End Sub

    Private Sub TextBox1_TextChanged(sender As Object, e As EventArgs) Handles TextBox1.TextChanged

    End Sub

    Private Sub PictureBox2_Click(sender As Object, e As EventArgs) Handles PictureBox2.Click

    End Sub
End Class
